### PR TITLE
Adding Feed to app

### PR DESCRIPTION
**Pull request by Yardengz**
Added a html page to the website that presents the Feed details .
Presenet the option of making a new Post.

changes:
add url in feed to connect for with  the HighTeach urls.
Added a copule of functions in the feed/models.py to help present the data in the page.
added form for the feed

image

link to issue:
https://github.com/redhat-beyond/HighTeach/issues/120

### CODE REVIEW

**Comment by Omeramsc** [comment]
Please squash commits, and make sure the commit left has a meaningful title and description.

File: feed/templates/feed_page.html [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        font-size: 16px;
+        margin-bottom: 10px;
+      }
+
+      .reply-content {
+        color: #444;
+        font-size: 14px;
+        margin-bottom: 6px;
+      }
+    </style>
```
Comment (shirazyomtov) [inline feed/templates/feed_page.html:95]: Please move the style to a css file and link it to the html file

**Review by shirazyomtov — CHANGES_REQUESTED** [review]
Great work I left a comment

File: feed/forms.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        self.user_id = user
+        super().__init__(*args, **kwargs)
+
+    def save(self, commit=True):
+        new_post = super().save(commit=False)
+        new_post.user_id = self.user_id
+        if commit:
+            new_post.save()
+
+        # new_post.join_group(self.user)
```
[the remark this answers is not in the document]
Comment (Yardengz) [reply feed/forms.py]: done

File: feed/templates/create_form.html [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    {% block content %}
+  <h1>Create New Post</h1>
+  <form method="POST">
+    {% csrf_token %}
+    {{ form.as_p }}
+    <button type="submit">Create</button>
+  </form>
+{% endblock %}
+  </body>
+</html>
```
[the remark this answers is not in the document]
Comment (Yardengz) [reply feed/templates/create_form.html]: done

File: feed/views.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    def post(self, request):
+        user = request.user
+        form = FeedForm(request.POST, user_id=user)
+        form.save(commit=False)
+        if form.is_valid():
+            form.save()
+        return redirect('feed_list')
+
+# class PostCreateView(LoginRequiredMixin, CreateView):
```
[the remark this answers is not in the document]
Comment (Yardengz) [reply feed/views.py]: done

File: tests/feed/test_feed.py [hunk 5]
```diff
@@ -77,3 +77,29 @@ def test_posts_order_in_dict_by_creation_date(self, num_of_posts, persist_user,
 
         posts_for_user = list(PostManager.get_posts_for_user(persist_user).keys())
         assert all(earlier.date <= later.date for earlier, later in zip(posts_for_user, posts_for_user[1:]))
+
+
+@pytest.mark.django_db
+class TestPostsListView:
+    def test_posts_list_view_loaded_authorized(self, authorized_client):
+        post_list_response = authorized_client.get('/feed/list/')
+        assert post_list_response.status_code == 200
```
Comment (Omeramsc) [inline tests/feed/test_feed.py:86]: I'd like to see a test that check for the existence of an expected post.

File: tests/feed/test_feed.py [hunk 6]
```diff
@@ -77,3 +77,29 @@ def test_posts_order_in_dict_by_creation_date(self, num_of_posts, persist_user,
...
+
+    def test_Posts_list_view_loaded_unauthorized(self, client):
```
Comment (Omeramsc) [inline tests/feed/test_feed.py:88]: test name should include only lower letters

File: tests/feed/test_feed.py (hunk 5 above)
Comment (Yardengz) [reply tests/feed/test_feed.py:86]: Isnt it the last test here? I am creating a post and check that it exist

Comment (Omeramsc) [reply tests/feed/test_feed.py:86]: The last test seems to check that the Post object was created, I want you to check a post created actually appears in the posts view.

File: tests/feed/test_feed.py (hunk 6 above)
Comment (Yardengz) [reply tests/feed/test_feed.py]: done

File: tests/feed/test_feed.py (hunk 5 above)
Comment (Yardengz) [reply tests/feed/test_feed.py:86]: done, is that what you meant?

File: feed/templates/feed_page.html (hunk 1 above)
Comment (Yardengz) [reply feed/templates/feed_page.html]: done

**Comment by Yardengz** [comment]
everything done

File: tests/feed/test_feed.py [hunk 7]
```diff
@@ -77,3 +77,35 @@ def test_posts_order_in_dict_by_creation_date(self, num_of_posts, persist_user,
...
+        post_creation_response = client.post("/feed/list/", data=post_creation_form_args)
+
+        assert post_creation_response.status_code == 302
+
+        assert Post.objects.filter(**post_creation_form_args).exists()
+
+    def test_created_post_in_view(self, client, persist_post_for_first_course):
+        client.force_login(persist_post_for_first_course.user_id)
+        response = client.get('/feed/list/')
+        assert response.status_code == 200
```
Comment (Omeramsc) [inline tests/feed/test_feed.py:110]: those two lines already exist in `test_posts_list_view_loaded_authorized`

File: tests/feed/test_feed.py [hunk 8]
```diff
@@ -77,3 +77,35 @@ def test_posts_order_in_dict_by_creation_date(self, num_of_posts, persist_user,
...
+            "course_id": persist_course.course_id,
+            "user_id": persist_user,
+            "content": "yarden"
+        }
+
+        assert not Post.objects.filter(**post_creation_form_args).exists()
+        post_creation_response = client.post("/feed/list/", data=post_creation_form_args)
+
+        assert post_creation_response.status_code == 302
+
```
Comment (Omeramsc) [inline tests/feed/test_feed.py:104]: no need for a blank line here

File: tests/feed/test_feed.py [hunk 9]
```diff
@@ -77,3 +77,35 @@ def test_posts_order_in_dict_by_creation_date(self, num_of_posts, persist_user,
...
+        assert post_list_response.status_code == 302
+
+    def test_create_post(self, client, persist_course, persist_user, persist_first_student_course):
+        client.force_login(persist_user)
+        post_creation_form_args = {
+            "course_id": persist_course.course_id,
+            "user_id": persist_user,
+            "content": "yarden"
+        }
+
```
Comment (Omeramsc) [inline tests/feed/test_feed.py:99]: same

File: feed/templates/feed_page.html [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+          {% for reply in replies %}
+            <p class="reply-content">Reply: {{ reply.content }}</p>
+          {% endfor %}
+        </div>
+      {% endfor %}
+      <p class="EOM">(End of posts)</p>
+    </div>
+  </body>
+
+{% endblock %}
```
Comment (natifridman) [inline feed/templates/feed_page.html:34]: Missing newline at end of file

File: static/styles/feed.css [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        color: #333;
+        font-size: 16px;
+        margin-bottom: 10px;
+      }
+
+      .reply-content {
+        color: #444;
+        font-size: 14px;
+        margin-bottom: 6px;
+}
```
Comment (natifridman) [inline static/styles/feed.css:89]: same

File: tests/feed/test_feed.py [hunk 12]
```diff
@@ -77,3 +77,35 @@ def test_posts_order_in_dict_by_creation_date(self, num_of_posts, persist_user,
...
+class TestPostsListView:
+    def test_posts_list_view_loaded_authorized(self, authorized_client):
+        post_list_response = authorized_client.get('/feed/list/')
+        assert post_list_response.status_code == 200
+
+    def test_posts_list_view_loaded_unauthorized(self, client):
+        post_list_response = client.get('/feed/list/')
+        assert post_list_response.status_code == 302
+
+    def test_create_post(self, client, persist_course, persist_user, persist_first_student_course):
```
Comment (natifridman) [inline tests/feed/test_feed.py:92]: What is the difference between this and `test_created_post_in_view`? The names look similar

Comment (Yardengz) [reply tests/feed/test_feed.py:88]: removed duplicate test

File: feed/templates/feed_page.html (hunk 10 above)
Comment (Yardengz) [reply feed/templates/feed_page.html]: added

File: tests/feed/test_feed.py (hunk 9 above)
Comment (Yardengz) [reply tests/feed/test_feed.py]: removed

File: tests/feed/test_feed.py (hunk 8 above)
Comment (Yardengz) [reply tests/feed/test_feed.py]: removed

File: tests/feed/test_feed.py (hunk 7 above)
Comment (Yardengz) [reply tests/feed/test_feed.py:103]: removed duplicate test

File: tests/feed/test_feed.py (hunk 5 above)
Comment (Yardengz) [reply tests/feed/test_feed.py]: removed duplicate test

File: tests/feed/test_feed.py [hunk 13]
```diff
@@ -77,3 +77,24 @@ def test_posts_order_in_dict_by_creation_date(self, num_of_posts, persist_user,
...
+        assert not Post.objects.filter(**post_creation_form_args).exists()
+        post_creation_response = client.post("/feed/list/", data=post_creation_form_args)
+        assert post_creation_response.status_code == 302
+        assert Post.objects.filter(**post_creation_form_args).exists()
+
+    def test_created_post_in_view(self, client, persist_post_for_first_course):
+        client.force_login(persist_post_for_first_course.user_id)
+        response = client.get('/feed/list/')
+        assert response.status_code == 200
+        assert any(persist_post_for_first_course.post_id == post.post_id for post in response.context['posts'])
```
Comment (Omeramsc) [inline tests/feed/test_feed.py:100]: For readability sake, I'll prefer you declare the result of `any` as a variable before asserting it.

**Review by Omeramsc — CHANGES_REQUESTED** [review]
Left one minor comment, other than that LGTM

**Comment by natifridman** [comment]
Please link this PR to an issue

**Comment by natifridman** [comment]
@Yardengz any update on this?

**Comment by Yardengz** [comment]
what should i do ?

**Comment by natifridman** [comment]
> what should i do ?

@Yardengz 
- Resolve conflicts
- Make sure you fixed all the issues raised in the review
- Link this PR to an issue

File: tests/feed/test_feed.py (hunk 13 above)
Comment (Yardengz) [reply tests/feed/test_feed.py]: done

**Review by Omeramsc — APPROVED** [review]
LGTM

**Pull request merged**